pkg/api/pin.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+// pinChunk pin's the already created chunk given its address.
+// It fails if the chunk is not present in the local store.
+// It also increments a pin counter to keep track of how many
+// pin requests are originating for this chunk.
+func (s *server) pinChunk(w http.ResponseWriter, r *http.Request) {
+	addr, err := swarm.ParseHexAddress(mux.Vars(r)["address"])
+	if err != nil {
+		s.logger.Debugf("pin chunk: unable to parse chunk address: %v", err)
+		s.logger.Error("pin chunk: unable to parse address")
+		jsonhttp.BadRequest(w, "bad address")
+		return
+	}
+
+	switch err := s.storer.Set(r.Context(), storage.ModeSetPin, addr); {
+	case errors.Is(err, storage.ErrNotFound):
+		ch, err := s.storer.Get(r.Context(), storage.ModeGetRequest, addr)
+		if err != nil {
+			s.logger.Debugf("pin chunk: unable to get chunk for address %q from netstore: %v", addr, err)
+			s.logger.Error("pin chunk: unable to get chunk for address %q from netstore")
+			jsonhttp.NotFound(w, nil)
+			return
+		}
+
+		_, err = s.storer.Put(r.Context(), storage.ModePutRequestPin, ch)
+		if err != nil {
+			s.logger.Debugf("pin chunk: unable to put chunk for %q into the localstore: %v", addr, err)
+			s.logger.Error("pin chunk: unable to put chunk into the localstore")
+			jsonhttp.InternalServerError(w, http.StatusText(http.StatusInternalServerError))
+			return
+		}
+	case err != nil:
+		s.logger.Debugf("pin chunk: unable to pink chunk for %q: %v, address %s", addr, err)
+		s.logger.Error("pin chunk: unable to pin chunk")
+		jsonhttp.InternalServerError(w, http.StatusText(http.StatusInternalServerError))
+		return
+	}
+
+	// For simplicity, we don't handle the case when the creation of
+	// a tracking pin for the address fails after storing the chunks.
+	// The two operations should be atomic otherwise.
+	if err := s.pinning.CreatePin(addr); err != nil {
+		s.logger.Debugf("pin chunk: creation of pin for %q failed: %v", addr, err)
+		s.logger.Error("pin chunk: creation of pin failed")
+		jsonhttp.InternalServerError(w, nil)
+		return
+	}
+
+	jsonhttp.OK(w, nil)
+}
+
+// unpinChunk unpin's an already pinned chunk. If the chunk is
+// not present or if the pin counter is zero, it raises error.
+func (s *server) unpinChunk(w http.ResponseWriter, r *http.Request) {
```

Comment:
likewise. both should be handled with the traversal service. have a look at how the traversal service is used in `pin_bytes.go` and `pin_files.go`. we should be using it in a similar way, although, some minor changes to the traversal service might be needed to correctly accomplish the task